File: NP/time_server/time_client.py
import socket
import time
import unittest


class TimeClient:
    def __init__(self, host='localhost', port=5000):
        self.address = (host, port)
        self.__sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.__sock.connect(self.address)  
    def __del__(self):
        print("클라이언트 소켓이 닫힙니다.")
        self.__sock.close()

    def get_time(self):
        self.__sock.sendall(b'GET TIME')
        current_time = self.__sock.recv(1024)
        if not current_time:
            print("서버로부터 응답을 받지 못했습니다.")
            return
        print("현재시각: ", current_time.decode())


class TestTimeClient(unittest.TestCase):
    def test_get_time(self):
        client = TimeClient()
        for i in range(5):
            client.get_time()
            time.sleep(3)
            
        self.assertIsNotNone(client.__sock, "Socket should not be None after connection")
        client.__del__()

File: NP/time_server/TimeClient.java
import java.io.*;
import java.net.*;

public class TimeClient {
    private Socket socket;

    public TimeClient(String host, int port) throws IOException {
        socket = new Socket(host, port);
        System.out.println("서버에 연결되었습니다.");
    }

    public void getTime() {
        try {
            OutputStream out = socket.getOutputStream();
            InputStream in = socket.getInputStream();

            out.write("GET TIME".getBytes());
            out.flush();

            byte[] buffer = new byte[1024];
            int read = in.read(buffer);
            if (read == -1) {
                System.out.println("서버로부터 응답을 받지 못했습니다.");
                return;
            }
            String currentTime = new String(buffer, 0, read);
            System.out.println("현재시각: " + currentTime);
        } catch (IOException e) {
            System.out.println("오류: " + e.getMessage());
        }
    }

    public void close() {
        try {
            socket.close();
            System.out.println("클라이언트 소켓이 닫힙니다.");
        } catch (IOException e) {
            System.out.println("소켓 닫기 오류: " + e.getMessage());
        }
    }

    public static void main(String[] args) throws Exception {
        TimeClient client = new TimeClient("localhost", 5000);
        for (int i = 0; i < 5; i++) {
            client.getTime();
            Thread.sleep(3000); // 3초 대기
        }
        client.close();
    }
}
